### ISSUE TITLE
[BE] Evaluate and improve eager for-loop optimizer memory perf

### ISSUE THREAD

**Issue by janeyx99** [pytorch/pytorch#104789]
We are not the most careful with using the least amount of memory in our for-loop implementations of our optimizers. For example, we have `max_exp_avg_sqs[i].copy_(torch.maximum(max_exp_avg_sqs_i, exp_avg_sq))` in adam.py which could be `torch.maximum(max_exp_avg_sqs_i, exp_avg_sq, out = max_exp_avg_sqs[i])` to save an allocation. Motivated by this comment: https://github.com/pytorch/pytorch/pull/104781#issuecomment-1625621945

This issue is to track the task of going through each of our for-loop (single_tensor) optimizers and ensuring that we are clean and minimal with our memory footprint.
- [ ] Adam
- [ ] AdamW
- [ ] Adagrad
- [ ] Adadelta
- [ ] Adamax
- [ ] NAdam
- [ ] RAdam
- [ ] ASGD
- [ ] SGD
- [ ] Rprop
- [ ] RMSprop
- [ ] LBFGS
- [ ] SparseAdam

cc @vincentqb @jbschlosser @albanD

**Comment by rishabbala** [comment]
Hi @janeyx99 can you assign this PR to me?

**Comment by janeyx99** [comment]
@rishabbala done!

**Comment by Vic-41148** [comment]
Is this still being worked on? No PR in 2-3 years happy to pick this up if it's available

**Comment by janeyx99** [comment]
@Vic-41148 feel free to take a look; the idea here isn’t to open PRs right away but to do an analysis on all our optimizers to improve memory usage. There are sometimes tradeoffs between readability and performance so I’m looking for someone who’s willing to go in and understand the theoretical bounds and the practical constraints. It may turn out that all our for loop optimizers are already memory efficient (unlikely) and no code needs to be written except for the test case that ensures we don’t regress!

**Comment by Vic-41148** [comment]
Thank you for letting me know :] I'll take some time looking through it here one by one only getting into code where I think is needed, I'll try my best and let you know as well if I make any progress!

**Comment by janeyx99** [comment]
Nice! Another angle to take this is from the higher level view--e.g., maybe we should be able to expect that peak memory will not be more than O(param + grad + state + one intermediate). So one can start with the test that encapsulates that, and then see what fails.

**Comment by Vic-41148** [comment]
Started working on this one, I did run the budget test across all 17 for loop optimizer/flag combos on main all 17 already do fit,
the caching allocator reuses freed blocks so short lived temp don't show up in peak even when code allocates them

I am now going through the _single_tensor_* function for spots where we can do write in-place instead, a handful of them in Adam/Rprop/NAdam capturable paths and a separate one in RAdam's update chain but that's a bigger change so keeping it separated for now, None of these move the measured peak though, but they do bring capturable branches in line with the non capturable and help where allocator can't reuse them

Still working on the verification of this so no pr yet (for now) also can I have this one assigned to me

**Comment by janeyx99** [comment]
Yes, the verification + built upon understanding + documentation is ultimately more important/will help guide decisions

**Comment by Vic-41148** [comment]
### A little update 

I went on to the verification side of things, I did budget test confirm it ran param + grad + state + 1 intermediate
on all 17 optimizers on main (used opt.state introspection for that, a couple of them have scalar enteries that aren't param-sized like NAdam mu_product), all 17 already do fit within budget, as you expected catching allocator is  freeing blocks and reusing it well so no short lived temporaries even show up in peak 

I did go through _single_tensor_* line by line here across all the 13 optimizers, 7 spots there can write in place instead of allocating, mostly in the Adam/Rprop/NAdam capturable paths and a diff  one in RAdam update chain 

None of these move peak memory though, but do bring capturable branches in line with non capturable branches and can help where allocator can't reuse (CUDA graph capture, first-iteration allocation), all these changes were gated on differentiable=False so autograd paths are untouched 

Split this into two branches 6 low risk changes for Adam/AdamW, NAdam, Rprop and theRAdam update chain separate it's a bit bigger of a structural change 

Also worked out a test suite for this got 89 total tests between those 2 covers gradcheck, dtype state_dict round-trips, torch.compile compatibility, and CUDA graph capture+replay compared against eager, had a few bugs there in the tests themselves too on the way  an assertion that was always true by construction, and a reference test that was checking its own algebra instead of the actual shipped function oops

Both are committed locally for now and lint clean, held off on the PR for now, Happy to hear if this was useful and if I should open a PR or keep going in this direction

**Comment by janeyx99** [comment]
I was following but I am not sure why you need to add that many more test cases. Are the ones we have today + the one that guarantees we're within a theoretical memory budget not enough?

**Comment by Vic-41148** [comment]
yes um I had been working on that, yeah you're right a lot of the tests were covered generically by @optims(optim_db) and I decided to cut them out, still have kept the test files separate  test_optimizer_memory.py and test_radam_memory.py as they are specific to these changes, I did restructure them to match how test_optim.py is written tho with loops and config lists 

After dropping the redundant parts and collapsing the rest (the memory budget checks were one test method per dtype/optimizer thats's why we ballooned up oops that why those are now loops and config lists) 

I got to about 39 tests from the original 89,I did commit locally again and all tests pass, lint is clean, and torch/ stays the same no changes there

For now It feels better than before but I'd like to hear back from you, let me know if I should open a PR

**Comment by janeyx99** [comment]
Let's move all tests to test_optim.py. If we want to restructure we should have an optim/test_optim_memory.py that contains all the memory tests, but that is an orthogonal refactor.

I am still confused why we have 39 new tests. Why is there special casing for any optimizer? Shouldn't we just have 1 memory budget test that is parametrized across all forloop optimizers? Or do you mean there are 39 parametrizations of 1 test? And if there no changes in torch, is this a test only PR?

**Comment by Vic-41148** [comment]
No No not a test only PR I do have changes in torch/, I have in place rewrites in adam, nadam, rprop, radam _single_tensor_* function
it's the same thing but fewer allocations where autograd doesn't need the intermediate 

The count is down once again to 26 of them now (20 on adam,nadam,rprop, 6 on radam here) 

But uh still wip for these:

- still hand written methods with a loop or config list inside, not yet @optims(optim_db) decorator parametrization (wip), so it's not 39 parametrizations of 1 test it's more like a few methods each looping internally, I will update once I have worked this out

- tests are still in their own files for now, will move into test_optim.py soon 

Once done with those I think I am good to open PRs, or if you want I can open them as is for now if you prefer to see where we are and go from there

**Comment by janeyx99** [comment]
Ok, cool. One thing you should try is to make the memory spike visible, like having a test with one big param. Otherwise if the test cannot catch anything, it is easy to regress.

**Comment by Vic-41148** [comment]
#### Update

Tests are now in test_optim.py  using  @optims(optim_db)  parametrization confirmed that each optimizer generates its own reported instance not a loop in a method now, all standalone files removed too

Also um when I said before that all of them were within budget before, I actually wasn't measuring the changes, CUDA defaults foreach defaults to True  if not specified so that was not correct (oops) wrong kernel , after forcing it to false both Adam and NAdam do eliminate one param sized allocation each step here

Rprop is diff tho no alloc reduction here, rewriting ops on already allocated tensors, regression test woulf be an oracle equivalence test

Also about bumping N, that alone didn't work, so I checked the budget formula and it had state bytes allocated before the measurement started, that creates a slack that grows with N, so reverted version would pass any size, fixed with , N independent assertion here scoped to the config where  rewrite will affect peak byte,I tested it and the pre patch ver failed at 11,264 bytes vs 8,192-bytes, this patch passed 7,680 bytes

I am opening the adam, nadam, rprop PR in a bit, I did test it and the diff was clean with current main, tests pass too, the other PR I'm still on wip will inform how that goes

**Comment by Vic-41148** [comment]
A correction to one of my last updates actually NAdam's peak mem didn't change, I thought going from .sqrt() to .sqrt_() removing a temporary would drop it same as Adam but :[ not really, I measured it while working and it's 14,848 bytes before and after, no real change

Only Adam shows a drop from 11,264 to  7,680 bytes, I suppose the change is still worth keeping because of the fewer allocations, but not moving peak memory, I will update the PR to match this as well

**Comment by Vic-41148** [comment]
## Update 

#### I have opened both the PR's now with the optimizations now 

### In summary:
  
1.Adam/NAdam/Rprop PR #194971:
 - replaced out-of-place temps with in-place operations in _single_tensor_adam
 - Reduces peak memory usage for Adam/AdamW from 11,264 bytes to 7,680 bytes (by saving 1 parameter-sized intermediate)
 - Streamlines allocation counts for NAdam/Rprop (peak remains unchanged due to other internal structures)

2.RAdam #195886:
 - Rewrote the capturable=False, rho_t > 5.0 update path in _single_tensor_radam to use in-place operations
 - Reduced peak memory usage in this from 20,480 bytes to 16,384 bytes (saving 1 parameter-sized intermediate here as well)
 - Keeping the capturable=True path regression-free for torch.compile compatibility

Both PRs have exact memory assertion unit tests in test_optim.py to protect from any future regressions
